src/util/convertHTMLtoJSON.ts

```diff
@@ -0,0 +1,161 @@
+import _ from 'lodash'
+import { Element, HimalayaNode, Text, parse } from 'himalaya'
+import { Block } from '../action-creators/importText'
+
+interface PreBlock extends Block {
+  wasSpan?: boolean,
+  hasSpaces?: boolean,
+}
+
+/** Parses input HTML and saves in JSON array using Himalaya. */
+export const convertHTMLtoJSON = (html: string) => {
+
+  /** Check if string contains only characters from given 'includes' array. */
+  const isStringIncludesOnly = (str: string, includes: string[]) => {
+    const matched =
+      includes
+        .map(char => str.replace(new RegExp(`[^${char}]`, 'g'), '').length)
+        .reduce((acc, cur) => acc + cur, 0)
+    return matched === str.length
+  }
+
+  /** Retrieve attribute from Element node by key. */
+  const getAttribute = (key: string, node: Element) => {
+    const { attributes } = node
+    const attribute = attributes.find(attr => attr.key === key)
+    return attribute ? attribute.value : undefined
+  }
+
+  /** Removes empty nodes and comments from himalaya's JSON output. */
+  const removeEmptyNodesAndComments = (nodes: HimalayaNode[]) => {
+    return nodes.filter(node => {
+      if (node.type === 'element') {
+        if (node.tagName === 'br') return false
+        node.children = removeEmptyNodesAndComments(node.children)
+        return true
+      }
+      if (node.type === 'comment') return false
+      // remove text nodes containing only empty space and new line characters, such nodes could be created by himalaya in case of multi-line html input
+      if (node.type === 'text' && isStringIncludesOnly(node.content, ['\n', ' '])) return false
+      return node.content.length
+    }) as (Element | Text)[]
+  }
+
+  /** Append children to parent as children property if it's necessary. */
+  const joinChildren = (nodes: (PreBlock[] | PreBlock)[]): PreBlock[] | PreBlock => {
+    // in case of element with span tag around text (e.g. one <span>and</span> two)
+    if (nodes.some(node => !Array.isArray(node) ? node.wasSpan || node.hasSpaces : false)) {
+      // take all text elements
+      const splittedText = _.takeWhile(nodes, node => !Array.isArray(node))
+      // join their content in a single line
+      const fullScope = splittedText.map(node => node && !Array.isArray(node) ? node.scope : '').join('')
+      const children = _.dropWhile(nodes, node => !Array.isArray(node)).flat()
+      return {
+        scope: fullScope,
+        children,
+      } as PreBlock
+    }
+    // WorkFlowy import with notes
+    if (nodes.some(node => node && !Array.isArray(node) && node.scope === '=note')) {
+      const parent = _.first(nodes)
+      const children = _.tail(nodes)
+      return Object.assign({}, parent, { children: children.flat() }) as PreBlock
+    }
+    if (nodes.every(node => Array.isArray(node))) return nodes.flat()
+    if (nodes.some(node => Array.isArray(node))) {
```

Comment:
By the 4th `if` statement, the nodes have been iterated over up to 4 times. See if you can use a single loop to determine the appropriate return value, i.e. Refactor `joinChildren` to be `O(n)` instead of `O(4n)`.

Reply:
We can use `for` loop for this. It doesn't look elegant, but performance is better.